### DISCUSSION TITLE
Deploying microcks as stack

### DISCUSSION THREAD

**Discussion by 30ThreeDegrees** [microcks/microcks#740] [Q&A] [answered]
Hi all,

I was wanting to setup microcks on a small vm so myself and a friend could both call the same microcks server to save having to different interface configs between machines, so I thought I would deploy microcks as a stack on a ubuntu 22 machine with a fresh install of docker instead of using compose so I had persistent data and auto reboot.

When I first tried to deploy as a stack I received an  error about docker-compose compatibility as stack doesn't support version '2' anymore and I also converter the docker-compose into version '3', I also changed the ~tmp file storage to use a volume so I had persistent data.

Everything seems to start up fine but I try to access the login screen it just hangs and I see some keycloak errors, unfortunately I dont understand enough about keycloak so I thought I would ask the question after not finding much on how to deploy this as a stack from google, stack overflow or the docs.

Any help or suggestions on how to launch this as a stack or point me towards the relevant docs would be appreciated.

`
version: '3'

services:
  mongo:
    image: mongo:3.4.23
    container_name: microcks-db
    volumes:
      - microcks-mongo-data:/data/db

  keycloak:
    image: jboss/keycloak:14.0.0
    container_name: microcks-sso
    ports:
      - "18080:8080"
    environment:
      DB_VENDOR: "h2"
      KEYCLOAK_USER: "admin"
      KEYCLOAK_PASSWORD: "admin"
      KEYCLOAK_IMPORT: "/tmp/microcks-realm.json"
      KEYCLOAK_FRONTEND_URL: "http://localhost:18080/auth"
    volumes: 
      - microcks-keycloak-data:/tmp/microcks-realm.json

  postman:
    image: quay.io/microcks/microcks-postman-runtime:latest
    container_name: microcks-postman-runtime

  app:
    depends_on:
      - mongo
      - keycloak
      - postman
    image: quay.io/microcks/microcks:latest
    container_name: microcks
    ports:
      - "8080:8080"
      - "9090:9090"
    environment:
      - SPRING_PROFILES_ACTIVE=prod
      - SPRING_DATA_MONGODB_URI=mongodb://mongo:27017
      - SPRING_DATA_MONGODB_DATABASE=microcks
      - POSTMAN_RUNNER_URL=http://postman:3000
      - TEST_CALLBACK_URL=http://microcks:8080
      - SERVICES_UPDATE_INTERVAL=0 0 0/2 * * *
      - KEYCLOAK_URL=http://keycloak:8080/auth
      - KEYCLOAK_PUBLIC_URL=http://localhost:18080/auth
      #- MAX_UPLOAD_FILE_SIZE=3MB
      #
volumes:
  microcks-mongo-data:
  microcks-keycloak-data:
`

**Comment by lbroudoux** [comment, marked as answer]
Hello there,

Sorry for delay in answer. As I think you'll reach out the vm using its DNS name, you'll have to customize the `KEYCLOAK_PUBLIC_URL` so that it reflects the DNS name used. Once Keycloak can be reach, you'll also need to configure the `redirect_uri` for Microcks realm (default is `http://localhost:8080`) to match your vm DNS name as well. For that: connect to the Keycloak admin console, login using admin user, change realm to Microcks and add this new redirect url on the `microcks-app-js` client configuration.

Hope it helps!
